### ISSUE TITLE
Module version protection

### ISSUE THREAD

**Issue by andrewchambers** [janet-lang/janet#83]
(import foo) should probably not be able to import a janet library if the version differs 'too much', whatever that means. Before a stable release, I would probably say this means  minor versions are not compatible.

Janet should probably also reject libraries compiled with different incompatible configurations (like nanbox settings).

My main motivation for this is so my rust crate can package the amalgamation headers, but provide some protection against a version mismatch with the installed system janet.

**Comment by bakpakin** [comment]
We could try to embed the version and common settings in the shared object via the JANET_MODULE_ENTRY macro. This way, the module loader function `(native)` could check these symbols and see if they match the current symbols.

**Comment by andrewchambers** [comment]
Trying to work out how the MODULE_ENTRY macro should look:

Was thinking something semi hacky like:
```
#define JANET_MODULE_ENTRY(X) void _janet_init(X) { assert(janet_api_compatible()) {
```
And reworking the existing C extensions slightly.

**Comment by bakpakin** [comment]
Addressed in latest master (after c8cf7c24453609a530c2a7e447e8a32315736c80). This should be very robust and give detailed messages about the differences in build settings.
